test/test_pipeline.py

```diff
@@ -194,6 +194,113 @@ def test_get_config_creates_two_different_dependent_components_of_same_type():
+
+
+def test_get_config_creates_different_components_if_instances_differ():
+    child_a = ChildComponent()
+    child_b = ChildComponent()
+    child_c = ChildComponent()
+    parent = ParentComponent(dependent=child_a)
+    parent2 = ParentComponent(dependent=child_b)
+    p_ensemble = Pipeline()
+    p_ensemble.add_node(component=parent, name="ParentA", inputs=["Query"])
+    p_ensemble.add_node(component=parent2, name="ParentB", inputs=["Query"])
+    p_ensemble.add_node(component=child_c, name="Child", inputs=["Query"])
+
+    expected_components = [
+        {"name": "ParentA", "type": "ParentComponent", "params": {"dependent": "ChildComponent"}},
+        {"name": "ChildComponent", "type": "ChildComponent", "params": {}},
+        {"name": "ParentB", "type": "ParentComponent", "params": {"dependent": "ChildComponent_2"}},
+        {"name": "ChildComponent_2", "type": "ChildComponent", "params": {}},
+        {"name": "Child", "type": "ChildComponent", "params": {}},
+    ]
+
+    expected_pipelines = [
+        {
+            "name": "query",
+            "nodes": [
+                {"name": "ParentA", "inputs": ["Query"]},
+                {"name": "ParentB", "inputs": ["Query"]},
+                {"name": "Child", "inputs": ["Query"]},
+            ],
+        }
+    ]
+
+    config = p_ensemble.get_config()
+    for expected_pipeline in expected_pipelines:
+        assert expected_pipeline in config["pipelines"]
+    for expected_component in expected_components:
+        assert expected_component in config["components"]
+
+
+def test_get_config_reuses_same_unnamed_dependent_components():
+    child = ChildComponent()
+    parent = ParentComponent(dependent=child)
+    parent2 = ParentComponent(dependent=child)
+    p_ensemble = Pipeline()
+    p_ensemble.add_node(component=parent, name="ParentA", inputs=["Query"])
+    p_ensemble.add_node(component=parent2, name="ParentB", inputs=["Query"])
+
+    expected_components = [
+        {"name": "ParentA", "type": "ParentComponent", "params": {"dependent": "ChildComponent"}},
+        {"name": "ChildComponent", "type": "ChildComponent", "params": {}},
+        {"name": "ParentB", "type": "ParentComponent", "params": {"dependent": "ChildComponent"}},
+    ]
+
+    expected_pipelines = [
+        {"name": "query", "nodes": [{"name": "ParentA", "inputs": ["Query"]}, {"name": "ParentB", "inputs": ["Query"]}]}
+    ]
+
+    config = p_ensemble.get_config()
+    for expected_pipeline in expected_pipelines:
+        assert expected_pipeline in config["pipelines"]
+    for expected_component in expected_components:
+        assert expected_component in config["components"]
+
+
+def test_get_config_multi_level_dependencies():
+    child = ChildComponent()
+    intermediate = ParentComponent(dependent=child)
+    parent = ParentComponent(dependent=intermediate)
+    p_ensemble = Pipeline()
+    p_ensemble.add_node(component=parent, name="Parent", inputs=["Query"])
```

Comment:
Ok now, what happens if I call this node `ParentComponent`? Or if I call two nodes with the same name explicitly? I didn't see a check for this case anywhere.

Reply:
Good you spotted this. Here was actually a bug as the name of the node to be added wasn't known to the graph when running `_get_all_component_names()`: fixed that.